### DISCUSSION TITLE
Trend arrow output via nightscout api -> data feed from xDrip+ REST-API

### DISCUSSION THREAD

**Discussion by emp-00** [NightscoutFoundation/xDrip#2372] [Q&A]
I using nightscout version 14.2.5 and Dexcom CGM data is uploaded via xDrip+ 2022-06-20 REST-API. This issue is about the "trend arrow" which I'm accessing via the nightscout api in a desktop gadget, as an example here "-->":

![image](https://user-images.githubusercontent.com/60714349/194715408-337e1dff-304a-4247-9c98-5ce2f7774ccd.png)

As you can see in this case +12 mg/dl but the trend arrow is "**Flat**" (-->) - do you know why? This is 100% sure the output coming from the api which actually returns "**Flat**" most of the time. Very rarely it gives a "**45up/down**" and in my case I have never ever seen a "**doubleup/down**" arrow. Do you know why? Even looking at the moving average a "Flat" arrow does not make sense at all in this concrete example since I was coming from a "LOW" bg and going straight up in the last 4 readings, due to eating glucose ;-). T**he xDrip+ trend arrow is working fine always showing correct trend arrows**, I think xDrip+ uses a "last-3-value-moving-average" for the trend arrow. But why is the nightscout api delivering such strange, mostly "Flat", output? Check this out:

Next reading -> still "Flat" despite another +7 mg/dl ...
![image](https://user-images.githubusercontent.com/60714349/194715243-a5be8b2f-6618-482f-abbf-210a1ea1a4a0.png)

Proving it's not a bug in my gadget, here's the api output from the same timestamp:
![image](https://user-images.githubusercontent.com/60714349/194715266-57d3c4ca-eb6d-4372-8efb-86c01eee3506.png)

Here's the nightscout trend, bg going straight up in the last ~30 minutes but always getting just a "Flat" trend arrow api output:
![image](https://user-images.githubusercontent.com/60714349/194715260-65507a08-6ce3-45ca-813d-98f3dedc0cf8.png)

Question: **Is xDrip+ resposible for generating the nightscout api output** (xDrip REST API uploading the "Flat" etc strings) or is this calculated in the nightscout server-side code? Would this be an xDrip+ or nightscout issue? Any ideas how to fix this? Many thanks!

**Comment by emp-00** [comment]
According to "CGM in the Cloud" this is an xDrip+ problem, shall I open an issue for this? See the current trend arrow in xDrip+ (correct) and nightscout (wrong, again "Flat"). Seems like xDrip+ is uploading the "Flat". Maybe I need to change an xDrip+ setting? Any ideas? Many thanks.

![image](https://user-images.githubusercontent.com/60714349/194765595-264d3093-2019-496d-9cbc-ad43d02f333d.png)
![image](https://user-images.githubusercontent.com/60714349/194764142-599eafdc-7f84-4039-a6f4-01b8cf5e44a3.png)
![image](https://user-images.githubusercontent.com/60714349/194764266-07a96160-11d9-438c-a253-262f778a35a6.png)

**Comment by Navid200** [comment]
Have you seen this?
https://github.com/NightscoutFoundation/xDrip/issues/920

**Reply by emp-00** [reply]
That's an issue with respect to the "delta" values. Those are correct in my case - xDrip uploaded delta values are shown correctly in nightscout, see above screenshot (+5 mg/dl). My issue is the "**trend arrow**" or "direction" in nightscout-terms - as shown above, nightscout shows a "Flat" arrow most of the time and according to Ben West this is caused by the uploading device - in my case xDrip+. Maybe I need to change an xDrip+ setting or shall I create an issue?

**Reply by Navid200** [reply]
No, you should not create an issue unless it is requested clearly here.

**Reply by Navid200** [reply]
Can you disable "Show Graphical Trend Arrow" on the home shelf?
https://navid200.github.io/xDrip/docs/Display/TrendArrow

Will the discrepancy between xDrip and Nightscout remain?

**Reply by emp-00** [reply]
Good hint @Navid200 , I am using the "Graphical Trend arrow" (I like it, would like to keep!) - ok, have now switched it off for testing the effect. Of course for the current value it changed from "slightly up" to "Flat" which  OK for the current trend. Will need to wait for an ascending/descending trend and will let you know what happens.

**Reply by Navid200** [reply]
Actually, I see the same thing you have described even with the graphical trend arow disabled.
I will ask for advice.

**Reply by emp-00** [reply]
Many thanks @Navid200 for confirming this! I know I had this issue probably "always" - but now I became aware, because I programmed a desktop gadget showing the current BG + delta-value and the trend arrow. I never paid much attention to the arrow in the nightscout web-view, but now I clearly see that the arrow does not make sense... Even if xDrip+ would be using a last 5-value moving average (should be last 2 only I think?) there should be a more significant "arrow angle". Right now the uploaded arrow is practically useless...

**Reply by emp-00** [reply]
Here's proof! Now as promised with "graphical trend arrow" disabled. Just got a +14 mg/dl bg increase due to eating stuff and fact is: xDrip+ shows the correct "Full Up" arrow BUT obviously xDrip feeds nightscout with the wrong "Flat" notification!

![image](https://user-images.githubusercontent.com/60714349/194770588-e5af908a-5bee-403c-99bd-d577fe386a1d.png)

See also this reply from "CGM in the cloud":
![image](https://user-images.githubusercontent.com/60714349/194771169-dbf3430a-451c-4de5-a623-11ab55cf5692.png)

**Comment by tolot27** [comment]
It looks like it is a bug in xDrip NS uploader. Please can you confirm that xDrip gets the data directly from the Dexcom (is it a G5 or G6) and uploads it to NS directly (not by a follower).

**Reply by emp-00** [reply]
Yes, I 100% confirmed that xDrip+ gets the data directly from a Dexcom G6 (native mode) and uploads directly via REST API to nightscout. I don't have any followers. Does this help? Do you already have an idea how to fix this bug? In case I need to open an issue or if you need additional information just let me know.

**Reply by emp-00** [reply]
@tolot27 : Thanks for looking into this. Any news? Haven't heard back.

**Reply by emp-00** [reply]
@tolot27 / @Navid200 - may I now open an [official issue](https://github.com/NightscoutFoundation/xDrip/issues) on the above described problem?

**Reply by Navid200** [reply]
Only if @tolot27 approves and note before please.

  I still don't know if xDrip uploads something specific for the arrow.  And why
If xDrip already uploads the delta, why would it also upload a value representing the slope of the arrow, which should practically be the same concept already embedded in the delta?

**Reply by emp-00** [reply]
I was asking myself the same questions ... but see above, apparently xDrip is uploading not only the delta but also the state of the "arrow" as text/string. According to "CGM in the Cloud" my described issue is not caused by nightscout but by xDrip - apparently nightscout is only a database and xDrip is uploading all the relevant data via REST-API ... I agree this is not really efficient but probably there are historic good reasons behind. Anyway, since I'm relying on the nightscout API output I'd like to get this fixed. 

Ok, I will wait a little longer for @tolot27 to give feedback...

**Reply by Navid200** [reply]
>   apparently nightscout is only a database and xDrip is uploading all the relevant data via REST-API ... 

I don't think that is correct in all circumstances.  Neither should it be.
As an example think about the statistics.
xDrip calculates all of them including PGS, GVI, estimated HbA1c, Mean BG, Median BG.
Nightscout calculates all of those as well, and it should.
    
Nightscout runs a on a server.  xDrip runs on a mobile.  
I see no reason why Nightscout would need processing support from a mobile device.  
If xDrip can calculate something from the BG readings, so can Nightscout easily.

**Reply by emp-00** [reply]
We both fully agree here - but it's not the question "what could/should be": I got clear feedback that the nightscout API returns the "delta" and "direction" values uploaded by the remote CGM software --> In this case this is xDrip+

**Reply by Navid200** [reply]
@tzachi-dar Does xDrip upload a value other than delta to represent the arrow?
If yes, why?

**Reply by tolot27** [reply]
@Navid200 It also sends the direction. Look at https://github.com/NightscoutFoundation/xDrip/blob/master/app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/NightscoutUploader.java#L623.
@emp-00 Please can you report the status of the preference "cloud_storage_api_use_best_glucose". If "cloud_storage_api" is not enabled, it should be false.

**Reply by emp-00** [reply]
@tolot27 : Sorry, but I cannot find both "cloud_storage_api_use_best_glucose" and "cloud_storage_api" in my xDrip+ settings. Can you please specifiy where to find these two options? Google search indicates that this is something in the xDrip+ source code, please help me there.

**Reply by emp-00** [reply]
@tolot27 : I'd love to report the status as per your question but please give me a hint "how to", see above. Thanks.
